spec/forms/user_piv_cac_setup_form_spec.rb

```diff
@@ -0,0 +1,138 @@
+        let(:other_user) { create(:user, :with_piv_or_cac) }
+        let(:x509_dn_uuid) { other_user.x509_dn_uuid }
+
+        it 'returns FormResponse with success: false' do
+          result = instance_double(FormResponse)
+
+          expect(FormResponse).to receive(:new).
+            with(success: false, errors: {}).and_return(result)
+          expect(Event).to_not receive(:create)
+          expect(form.submit).to eq result
+          expect(user.reload.piv_cac_enabled?).to eq false
+          expect(form.error_type).to eq 'piv_cac.already_associated'
+        end
+
+        context 'when we encounter a race condition between checking and storing' do
+          let(:x509_dn_uuid) { other_user.x509_dn_uuid + 'X' }
+
+          it 'returns FormResponse with success: false' do
+            allow(user).to receive(:save!).and_raise(PG::UniqueViolation)
+
+            result = instance_double(FormResponse)
+
+            expect(FormResponse).to receive(:new).
+              with(success: false, errors: {}).and_return(result)
+            expect(Event).to_not receive(:create)
+            expect(form.submit).to eq result
+            expect(user.reload.piv_cac_enabled?).to eq false
+            expect(form.error_type).to eq 'piv_cac.already_associated'
+          end
+        end
+      end
+    end
+
+    context 'when token is invalid' do
+      let(:token) { 'bad-token' }
+      let(:token_response) do
+        { 'error' => 'token.bad', 'nonce' => nonce }
+      end
+
+      it 'returns FormResponse with success: false' do
+        result = instance_double(FormResponse)
+
+        expect(FormResponse).to receive(:new).
+          with(success: false, errors: {}).and_return(result)
+        expect(Event).to_not receive(:create)
+        expect(form.submit).to eq result
+        expect(user.reload.piv_cac_enabled?).to eq false
+        expect(form.error_type).to eq 'token.bad'
+      end
+    end
+
+    context 'when nonce is invalid' do
+      let(:token) { 'bad-token' }
+      let(:token_response) do
+        { 'error' => 'token.bad', 'nonce' => bad_nonce }
+      end
+      let(:bad_nonce) { nonce + 'X' }
+
+      it 'returns FormResponse with success: false' do
+        result = instance_double(FormResponse)
+
+        expect(FormResponse).to receive(:new).
+          with(success: false, errors: {}).and_return(result)
+        expect(Event).to_not receive(:create)
+        expect(form.submit).to eq result
+        expect(form.error_type).to eq 'token.invalid'
+      end
+    end
+
+    context 'when token is missing' do
+      let(:token) { }
+
+      it 'returns FormResponse with success: false' do
+        result = instance_double(FormResponse)
+
+        expect(FormResponse).to receive(:new).
+          with(success: false, errors: {}).and_return(result)
+        expect(Event).to_not receive(:create)
+        expect(form.submit).to eq result
+        expect(user.reload.piv_cac_enabled?).to eq false
+      end
+    end
+  end
```

Comment:
In all of these invalid scenarios, we want the `errors` hash to contain the reason for the failure, which is why I suggested to use ActiveModel validations for all the different validations.